digdag-core/src/main/java/io/digdag/core/database/AESGCMSecretCrypto.java

```diff
@@ -0,0 +1,177 @@
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.Provider;
+import java.security.SecureRandom;
+import java.security.Security;
+import java.util.Base64;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static javax.crypto.Cipher.ENCRYPT_MODE;
+
+public class AESGCMSecretCrypto implements SecretCrypto
+{
+    private static final SecureRandom SECURE_RANDOM = new SecureRandom();
+
+    private final SecretKey sharedSecret;
+
+    private static final int AES_KEY_SIZE = 128;
+    private static final int GCM_NONCE_LENGTH = 12;
+    private static final int GCM_TAG_LENGTH = 16;
+
+    private static final int TERM = 1;
+    private static final int VERSION = 1;
+
+    private static final int MAX_PLAINTEXT_LENGTH = 1024;
+
+    private static final int RECORD_SIZE = 2048;
+    private static final int LENGTH_SIZE = 4;
+    private static final int TEXT_SIZE = RECORD_SIZE - LENGTH_SIZE;
+    private static final int TERM_SIZE = 4;
+    private static final int VERSION_SIZE = 4;
+    private static final int OPAQUE_SIZE = TERM_SIZE + VERSION_SIZE + GCM_NONCE_LENGTH + RECORD_SIZE + GCM_TAG_LENGTH;
+
+    public AESGCMSecretCrypto(String sharedSecretBase64)
+    {
+        byte[] sharedSecretRaw = Base64.getDecoder().decode(sharedSecretBase64);
+        Preconditions.checkArgument(sharedSecretRaw.length * 8 == AES_KEY_SIZE);
+        this.sharedSecret = new SecretKeySpec(sharedSecretRaw, "AES");
+    }
+
+    @Override
+    public String encryptSecret(String plainText)
+    {
+        if (plainText.length() > MAX_PLAINTEXT_LENGTH) {
+            throw new IllegalArgumentException("Too long text");
+        }
+
+        byte[] nonce = generateNonce();
+
+        Cipher cipher = cipher(ENCRYPT_MODE, sharedSecret, nonce);
+
+        byte[] plainTextBytes = plainText.getBytes(UTF_8);
+
+        if (plainTextBytes.length > MAX_PLAINTEXT_LENGTH) {
+            throw new IllegalArgumentException("Too long text");
+        }
+
+        byte[] recordBytes = new byte[RECORD_SIZE];
+        ByteBuffer recordBuffer = ByteBuffer.wrap(recordBytes);
+        recordBuffer.putInt(plainTextBytes.length);
+        recordBuffer.put(plainTextBytes);
+
+        byte[] cipherText;
+        try {
+            cipherText = cipher.doFinal(recordBytes);
+        }
+        catch (IllegalBlockSizeException | BadPaddingException e) {
+            throw Throwables.propagate(e);
+        }
+
+        assert cipherText.length == RECORD_SIZE + GCM_TAG_LENGTH;
+
+        byte[] opaque = new byte[OPAQUE_SIZE];
+        ByteBuffer output = ByteBuffer.wrap(opaque);
+
+        output.putInt(TERM);
+        output.putInt(VERSION);
```

Comment:
Note that `DisabledSecretCrypto` really means disabled, not clear-text. Any attempt to use secrets without explicitly setting `digdag.secret-encryption-key` will fail with an error.